README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# SDLfilter

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3631115.svg)](https://doi.org/10.5281/zenodo.3631115)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/SDLfilter)](https://cran.r-project.org/package=SDLfilter)

## Overview

SDLfilter contains a variety of functions to screen GPS/Argos locations
and to assess the adequacy of sample size of tracking data for animal
distribution analysis.

## Installation

``` r
# The official version from CRAN:
install.packages("SDLfilter")

# Or the development version from GitHub:
install.packages("devtools")
devtools::install_github("TakahiroShimada/SDLfilter")
```

## Usage

``` r
library(SDLfilter)
```

### 1. Location filtering

There are three main filtering functions.

-   *dupfilter* filters temporal and spatial duplicates.  
-   *ddfilter* filters locations with high error.  
-   *depthfilter* filters locations by water depth.

<!-- <p>&nbsp;</p> -->

#### 1-1. Load tracking data

``` r
data(turtle)
```

#### 1-2. Remove temporal and spatial duplicates

``` r
turtle.dup <- dupfilter(turtle)
```

#### 1-3. Remove biologically unrealistic fixes

``` r
## Calculate the maximum linear speed between two consecutive locations
V <- vmax(turtle.dup)  

## Calculate the maximum one-way linear speed of a loop trip
VLP <- vmaxlp(turtle.dup) 

## Run ddfilter
turtle.dd <- ddfilter(turtle.dup, vmax=V, vmaxlp=VLP)
```

#### 1-4. Plot data

<details>
<summary>
Click to show code
</summary>

``` r
 # Entire area
 p1 <- to_map(turtle.dup, bgmap=Australia, point.size = 2, line.size = 0.5, axes.lab.size = 0, 
             multiplot = FALSE, point.bg = "red",
             title.size=15, title="Entire area")[[1]] + 
   geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, fill="yellow", shape=21)+
   geom_point(aes(x=x, y=y), data=data.frame(x=c(154, 154), y=c(-22, -22.5)), 
              size=3, fill=c("yellow", "red"), shape=21) + 
   annotate("text", x=c(154.3, 154.3), y=c(-22, -22.5), label=c("Retained", "Removed"), 
            colour="black", size=4, hjust = 0)

 # Zoomed in
 p2 <- to_map(turtle.dup, bgmap=SandyStrait, xlim=c(152.7, 153.2), ylim=(c(-25.75, -25.24)), 
             axes.lab.size = 0, point.size = 2, point.bg = "red", line.size = 0.5, 
             multiplot = FALSE, title.size=15, title="Zoomed in")[[1]] + 
 geom_path(aes(x=lon, y=lat), data=turtle.dd, size=0.5, colour="black", linetype=1) + 
 geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, colour="black", shape=21, fill="yellow")
 
 ## plot
 gridExtra::grid.arrange(p1, p2, layout_matrix=cbind(1,2))
```

</details>

![](man/figures/unnamed-chunk-7-1.png)

### 2. Assessing sample sizes (probability-based approach)

#### 2-1. Input UDs

The input data can be either a matrix or a list of RasterLayer objects.
Each row of the matrix or each RasterLayer object contains the
probability distribution of an animal. The function assumes that each
column of a matrix is associated with a unique geographical location,
therefore it is critical that the grid size and geographical extent are
consistent across UDs. In this example (15 tracks of flatback turtles),
the grid size was 1km and the geographical extent was 1901789, 1972789,
-2750915, -2653915 (EPSG:3577) across all 15 layers.

<details>
<summary>
Click to show an example code of UD estimation.
</summary>

``` r
library(adehabitatHR); library(raster)

## Tracking data
data(flatback)
flatback <- track_param(flatback, param = "time") # calculate time between successive locations
flatback_id <- unique(flatback$id)


## Data range with 5km buffer
buff <- 5e+3
xmin <- min(flatback$x) - buff; xmax <- max(flatback$x) + buff
ymin <- min(flatback$y) - buff; ymax <- max(flatback$y) + buff

## Make a grid layer
cell.size <- 1e+3 # (1km x 1km)
x <- seq(xmin, xmax, cell.size)
y <- seq(ymin, ymax, cell.size)
xy.df <- expand.grid(x = x, y = y)
xy.coords <- SpatialPixels(SpatialPoints(xy.df))
xy.sp <- SpatialPoints(xy.coords, proj4string = CRS("+init=epsg:3577"))
z <- rep(1, nrow(xy.df))
xyz <- cbind(xy.df, z)
grid_spdf <- SpatialPixelsDataFrame(xy.coords, xyz)

## UD per turtle
ud_raster <- list()
for(i in 1:length(flatback_id)){
  
  ## ID
  ID <- flatback_id[i]

  ## Tracking data
  turtle.data <- with(flatback, flatback[id %in% ID, ])

  ## Creates an object of class Itraj
  data.ltraj <- with(turtle.data, as.ltraj(turtle.data[,c("x", "y")], date=DateTime, id=ID, burst=ID))
  
  ## Parameters for BRB
  TM = quantile(turtle.data$pTime, probs = 0.95, na.rm = TRUE)*3600
  LM = 50
  dp = BRB.likD(data.ltraj, Tmax=TM, Lmin=LM)
  HM = 100
  
  ## Estimate the UD
  ud <- BRB(data.ltraj, D=dp, Tmax=TM, Lmin=LM, hmin=HM, grid=grid_spdf, type="UD")
  
  ## Convert the UD to raster
  ud_raster[[i]] <- raster(ud)
}
```

</details>

``` r
# A matrix
data(ud_matrix)

# Or a list of RasterLayer/SpatRaster
data(ud_raster)
```

#### 2-2. Calculate overlap probability from 2000 random permutation (\> sample size x 100)

It will take some time to run this code depending on the number of
iterations and the machine specs. The runtime was about 2.5 minutes for
2000 iterations on a linux machine (Intel i7-4702HQ CPU @ 2.20GHz, 16GB
RAM).

``` r
overlap <- boot_overlap(ud_matrix, R = 2000, method = "PHR")
```

#### 2-3. Find the minimum sample size required to estimate the general distribution

An asymptote could be considered once the mean overlap probability
exceeded 95% of the estimated horizontal asymptote (Shimada et al 2021).
The sample size linked to this value is deemed to be the minimum sample
size required to represent the general distribution of the group.

``` r
a <- asymptote(overlap, upper.degree = 5, estimator = 'glm', family = binomial)
```

#### 2-4. Plot the estimated overlap probabilities relative to the sample sizes (black points), the 95% confidence intervals (grey lines), and 95% of the estimated asymptote (dashed line).

<details>
<summary>
Click to show code
</summary>

``` r
ggplot(data = a$results, aes(x = x))+
  geom_errorbar(aes(ymin = y_lwr, ymax = y_upr), width = 0.2, colour = 'darkgrey', size = 1) + 
  geom_point(aes(y = y), size = 2) + 
  geom_hline(yintercept = a$h.asymptote*0.95, linetype = 2) +
  scale_x_continuous(breaks = seq(0, 15, 3), limits = c(1.9,15.1), name = "Animals tracked (n)") +
  scale_y_continuous(limits = c(0.5,1), name = "Overlap probability") +
  theme_light()
```

</details>

![](man/figures/unnamed-chunk-13-1.png)

> Please see the package help pages and Shimada et al. (2012, 2016,
> 2021) for more details.

## References

If you use the function *ddfilter*, please cite

Shimada T, Jones R, Limpus C, Hamann M (2012) Improving data retention
and home range estimates by data-driven screening. *Mar Ecol Prog Ser*
457:171-180 doi:
[10.3354/meps09747](http://dx.doi.org/10.3354/meps09747)

If you use the functions *dupfilter* or *depthfilter*, please cite

Shimada T, Limpus C, Jones R, Hazel J, Groom R, Hamann M (2016) Sea
turtles return home after intentional displacement from coastal foraging
areas. *Mar Biol* 163:1-14 doi:
[10.1007/s00227-015-2771-0](http://dx.doi.org/10.1007/s00227-015-2771-0)

If you use the functions *asymptote*, *boot_overlap*, *combn_overlap* or
*boot_area*, please cite

Shimada, T, Thums, M, Hamann, M, et al. (2021) Optimising sample sizes
for animal distribution analysis using tracking data. *Methods Ecol
Evol* 12(2):288-297 doi:
[10.1111/2041-210X.13506](https://doi.org/10.1111/2041-210X.13506)

## Current version

Version 2.3.3 (2023-11-07)


NEWS
Version 2.3.3 (2023-11-07)

  o removed functions that relied on the archived plotKML package. The functions include kml_track and to_kml. 


Version 2.3.2 (2023-04-17)

  o fixed bugs in the ddfilter_speed when "method = 2" is selected.
  o significantly improved the processing speed for the ddfilter_speed, especially when "method = 1" is used.


Version 2.3.1 (2023-01-16)

  o added an additional argument "ia" in the distfilter. "ia" is the inner angle between consecutive locations that can be used for detecting potential outliers. Default (NA) is set to ignore this argument.
  o added an additional argument "nloc" in the vmaxlp function. This allows to specify the minimum number of successive locations to be considered a loop trip. The default value is set to be 5 locations.
  o fixed an error in the vmaxlp function that occurs when there is not enough data for the calculation.
  o fixed bugs in the "ML" method for the vmax and vmaxlp functions.
  

Version 2.3.0 (2022-10-11)

  o improved the estimation of vmax of vmaxlp by implementing Maximum Likelihood estimation.
  o fixed minor bugs in the ddfilter, dupfilter and track_param functions.
  o reinstated the kml_track function as the dependent package (plotKML) has been reinstated to the CRAN repository.
  o added a new function to_kmz, which generates KMZ from locations and track.
  

Version 2.2.1 (2022-05-15)

  o added a parallel computing option to the dupfilter_time, dupfilter_space, and dupfilter functions.
  o removed the kml_track function as the dependent package (plotKML) has been removed from the CRAN repository.
  o updated to use the sf and stars packages instead of the sp and raster packages. The updated functions are boot_area, boot_overlap, combn_overlap, depthfilter, dupfilter_space, track_param, dupfilter_time, and map_track
  o updated to use the dplyr package instead of the plyr package. The updated functions are boot_area, ddfilter_loop, ddfilter_speed, dupfilter_qi, dupfilter_space, dupfilter_time, and track_param.  
  o updated the track_param function to use the geosphere package instead of the trip package.
  o The map_track function has been replaced with the "to_map" function.


Version 2.2.0 (2022-01-10)

  o added an option in the asymptote function to estimate the confidence intervals for bootstrapped overlap estimates. The new default is to use the CIs when considering an asymptote.
  o added a new function 'combn_overlap' as an alternative to 'boot_overlap'. This new function calculates overlaps between all possible combination of UDs relative to sample size.
  o improved the processing speed of 'boot_overlap'.


Version 2.1.2 (2021-10-31)

  o fixed a minor bug related with the max.asymptote argument within the asymptote function.


Version 2.1.1 (2021-07-04)

  o amended the ddfilter, dupfilter, and their depending functions (e.g. ddfilter_speed) to return movement parameters used in each function.


Version 2.1.0 (2021-04-02)

  o improved the asymptote function.
  o added a new citation for functions boot_overlap, boot_area and asymptote. Shimada T, Thums M, Hamann M, Limpus CJ, Hays GC, FitzSimmons NN, Wildermann NE, Duarte CD, Meekan MG (2021) Optimising sample sizes for animal distribution analysis using tracking data. Methods in Ecology and Evolution, 12(2):288-297. doi:10.1111/2041-210X.13506 
  o updated the README file.
  o improved the R documentations of functions asymptote, boot_overlap, and boot_area.
  o renamed the example data "curtis" to "flatback".


Version 2.0.1 (2020-07-04)

  o fixed bugs in the track_param function. 
  o The input qi values can be now either the number of GPS satellites or Argos Location Classes. Argos Location Classes will be converted to numerical values, where "A", "B", "Z" will be replaced with "-1", "-2", "-3" respectively.
  o The input DateTime (GMT) can be now either in class POSIXct or character with the following format "2012-06-03 01:33:46".
  o improved the R documentations.


Version 2.0.0 (2020-05-04)

  o added four new functions (boot_overlap, boot_area, asymptote, percent_vol) to assess sample sizes of animal tracking data.
  o added a new function (track_param) to calculate some parameters of animal tracking data (i.e. time, distance, speed, angle between locations).
  o added a new function (kml_track) to visualise the tracking data.
  o added a new example data "curtis".
  o added a new argument "type" in the depthfilter function.
  o renamed some functions (ddfilter.loop -> ddfilter_loop; ddfilter.speed -> ddfilter_speed; dupfilter.exact -> dupfilter_exact; dupfilter.qi -> dupfilter_qi; dupfilter.space -> dupfilter_space; dupfilter.time -> dupfilter_time, distantfilter -> distfilter).
  o renamed an argument in ddfilter (maxvlp -> vmaxlp).
  o renamed the plotMap to map_track
  o improved the map_track function.
  o improved the R documentations.
  o improved the processing speed of the dupfilter function and its dependent functions.
  o updated the vmax and maxvlp functions. In the previous version, when input data contain duplicate locations, the function ceased with error messages. These errors are now avoided by removing duplicate locations using the dupfilter function prior to calculation of vmax and maxvlp.


Version 1.2.1 (2019-02-19)

  o amended "plotMap" in accordance with the ggsn update to version 0.5.0. 


Version 1.2.0 (2019-02-18)

  o added a mew function "distance_filter". This function removes fixes located beyond a given threshold distance.
  o fixed bugs in the example codes.
  o amended "plotMap" in accordance with the ggmap update to version 3.0.0. 


Version 1.1.1 (2018-06-10)


R/ddfilter.R
#' @aliases ddfilter
#' @title Filter locations using a data driven filter
#' @description Function to remove locations by a data driven filter as described in Shimada et al. (2012).
#' @param sdata A data frame containing columns with the following headers: "id", "DateTime", "lat", "lon", "qi". 
#' See the data \code{\link{turtle}} for an example.
#' The function filters the input data by a unique "id" (e.g. transmitter number, identifier for each animal). 
#' "DateTime" is the GMT date & time of each location in class \code{\link[base:DateTimeClasses]{POSIXct}} or 
#' \code{\link[base]{character}} with the following format "2012-06-03 01:33:46".
#' "lat" and "lon" are the latitude and longitude of each location in decimal degrees. 
#' "qi" is the quality index associated with each location fix. 
#' The input values can be either the number of GPS satellites or Argos Location Classes. 
#' Argos Location Classes will be converted to numerical values, where "A", "B", "Z" will be replaced with "-1", "-2", "-3" respectively.
#' The greater number indicates a higher accuracy. 
#' @param vmax A numeric value specifying a threshold of speed from a previous and/or to a subsequent fix. 
#' Default is 8.9km/h. 
#' If this value is unknown, it can be estimated from \emph{sdata} using the function \code{\link{vmax}}.
#' @param vmaxlp A numeric value specifying a threshold of speed, which is used to evaluate the locations of loop trips. 
#' Default is 1.8 km/h.
#' If this value is unknown, it can be estimated from \emph{sdata} using the function \code{\link{vmaxlp}}.
#' @param qi An integer specifying a threshold of quality index, which is used to evaluate the locations of loop trips. 
#' Default is 4.
#' @param ia An integer specifying a threshold of inner angle, which is used to evaluate the locations of loop trips. 
#' Default is 90 degrees.
#' @param method An integer (1 or 2) specifying how locations should be filtered with \emph{vmax}. 
#' Default is 1 (both way) and removes a location if the speed from a previous AND to a subsequent location exceeds \emph{vmax}. 
#' Select 2 (one way) to remove a location if the speed from a previous OR to a subsequent location exceeds \emph{vmax}. 
#' For the latter, the filter examines successive suspect locations (i.e. the speed from a previous and/or to a subsequent location exceeds \emph{vmax}) 
#' and retain one location that is associated with the minimum speed from a previous and/or to a subsequent location.
#' @importFrom gridExtra marrangeGrob
#' @export
#' @details Locations are removed if the speed from a previous and/or to a subsequent location exceeds \emph{vmax}, 
#' or if all of the following criteria apply: the associated quality index is less than or equal to \emph{qi}, 
#' the inner angle is less than or equal to \emph{ia} and the speed either from a previous 
#' or to a subsequent location exceeds \emph{vmaxlp}. 
#' If \emph{vmax} and \emph{vmaxlp} are unknown, they can be estimated using the functions \code{\link{vmax}} and \code{\link{vmaxlp}} respectively.
#' @return The input data is returned without locations identified by this filter. 
#' The following columns are added: "pTime", "sTime", "pDist", "sDist", "pSpeed", "sSpeed", "inAng". 
#' "pTime" and "sTime" are hours from a previous and to a subsequent fix respectively. 
#' "pDist" and "sDist" are straight distances in kilometres from a previous and to a subsequent fix respectively. 
#' "pSpeed" and "sSpeed" are linear speed from a previous and to a subsequent fix respectively. 
#' "inAng" is the degree between the bearings of lines joining successive location points.
#' @author Takahiro Shimada
#' @references Shimada T, Jones R, Limpus C, Hamann M (2012) 
#' Improving data retention and home range estimates by data-driven screening. 
#' \emph{Marine Ecology Progress Series} 457:171-180 \doi{10.3354/meps09747}
#' @seealso \code{\link{ddfilter_speed}}, \code{\link{ddfilter_loop}}, \code{\link{vmax}}, \code{\link{vmaxlp}}
#' @examples
#' #### Load data sets
#' ## Fastloc GPS data obtained from a green turtle
#' data(turtle)
#' 
#' ## A Map for the example site
#' data(Australia)
#' data(SandyStrait)
#' 
#' 
#' #### Filter temporal and/or spatial duplicates
#' turtle.dup <- dupfilter(turtle, step.time=5/60, step.dist=0.001)
#'  
#' 
#' #### ddfilter
#' ## Using the built-in function to estimate the threshold speeds
#' V <- vmax(turtle.dup)
#' VLP <- vmaxlp(turtle.dup)
#' turtle.dd <- ddfilter(turtle.dup, vmax=V, vmaxlp=VLP)
#' 
#' ## Or using user specified threshold speeds
#' turtle.dd <- ddfilter(turtle.dup, vmax=9.9, qi=4, ia=90, vmaxlp=2.0)
#' 
#' 
#' #### Plot data removed or retained by ddfilter
#' ## Entire area
#' p1 <- to_map(turtle.dup, bgmap=Australia, point.size = 2, line.size = 0.5, axes.lab.size = 0, 
#'             multiplot = FALSE, point.bg = "red",
#'             title.size=15, title="Entire area")[[1]] + 
#'   geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, fill="yellow", shape=21)+
#'   geom_point(aes(x=x, y=y), data=data.frame(x=c(154, 154), y=c(-22, -22.5)), 
#'              size=3, fill=c("yellow", "red"), shape=21) + 
#'   annotate("text", x=c(154.3, 154.3), y=c(-22, -22.5), label=c("Retained", "Removed"), 
#'            colour="black", size=4, hjust = 0)
#'
#' ## Zoomed in
#' p2 <- to_map(turtle.dup, bgmap=SandyStrait, xlim=c(152.7, 153.2), ylim=(c(-25.75, -25.24)), 
#'             axes.lab.size = 0, point.size = 2, point.bg = "red", line.size = 0.5, 
#'             multiplot = FALSE, title.size=15, title="Zoomed in")[[1]] + 
#' geom_path(aes(x=lon, y=lat), data=turtle.dd, linewidth=0.5, colour="black", linetype=1) + 
#' geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, colour="black", shape=21, fill="yellow")
#'
#' gridExtra::marrangeGrob(list(p1, p2), nrow=1, ncol=2)



ddfilter<-function(sdata, vmax=8.9, vmaxlp=1.8, qi=4, ia=90, method=1) {
  
  #### Sample size of the input data
  OriginalSS<-nrow(sdata)
 
  #### Run ddfilters
  cat("\n")
  sdata<-ddfilter_speed(sdata=sdata, vmax=vmax, method=method)
  sdata<-ddfilter_loop(sdata=sdata, qi=qi, ia=ia, vmaxlp=vmaxlp)
  
  
  #### Report the summary of filtering
  FilteredSS <- nrow(sdata)
  RemovedSamplesN <- OriginalSS - FilteredSS
  if(RemovedSamplesN > 0){
    RemovedSamplesP <- round((1-(FilteredSS/OriginalSS))*100,2)
  } else {
    RemovedSamplesP <- 0
  }
  
  
  cat("\n")
  cat("Input data:", OriginalSS, "locations", fill = TRUE)
  cat("Filtered data:", FilteredSS, "locations", fill = TRUE)
  cat("ddfilter removed ", RemovedSamplesN, " locations (", RemovedSamplesP, "% of original data)", sep="", fill = TRUE)
  cat("\n")
  
  #### Return the filtered data set
  return(sdata)
}
